### PR TITLE
Unified search results

### LINKED ISSUE #25816: Unified search search suggestions should based on which search engine is being used
Update what search suggestions are being shown by following the table below

Screen Shot 2022-06-29 at 12 55 26 PM

&nbsp;
**Petru: Confirmed with @Verdi that the "Ads" suggestions shown the table are not applicable.**
&nbsp;
┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20854)

### PR DESCRIPTION
**Pull request by Mugurell**
Needs https://github.com/mozilla-mobile/firefox-android/pull/138 & https://github.com/mozilla-mobile/firefox-android/pull/263

https://user-images.githubusercontent.com/11428869/208063432-8a834620-fd0a-44ad-b785-724d4067d6b7.mp4

~~Pending UX review by @Verdi.~~
Shared an APK for Michael Verdi to test and they approved the end result.

&nbsp;

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #25816

### CODE REVIEW

**Comment by Mugurell** [comment]
Tested with the changes from the AC tickets and everything should work nicely. Ready for review.

**Comment by Mugurell** [comment]
### **Request for data collection review**
1. What questions will you answer with this data?
To better understand the usefulness of awesomebar suggestions . 
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? 
This will allow to refine Awesomebar suggestions.
3. What alternative methods did you consider to answer these questions? Why were they not sufficient? 
No alternative is available to answer this question.
4. Can current instrumentation answer these questions? 
No current instrumentation can answer this question.
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement Description | Data Collection Category | Tracking Bug #
-- | -- | --
Awesomebar search term suggestion was clicked | category 2 -   interaction data | #25816

6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way. 
https://github.com/mozilla-mobile/fenix/blob/master/docs/metrics.md
7. How long will this data be collected? Choose one of the following: 
This will be set to never expire, similar to the other telemetry for similar search suggestion interactions.
8. What populations will you measure? 
No filters - all locales, channels, etc.
9. If this data collection is default on, what is the opt-out mechanism for users? 
Default Glean SDK opt-out mechanism.
10. Please provide a general description of how you will analyze this data. 
Glean / Amplitude
11. Where do you intend to share the results of your analysis? 
Only on Glean, Amplitude and with mobile teams.
12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
 No third-party tool will use this data.

File: app/metrics.yaml [hunk 1]
```diff
@@ -7654,6 +7654,22 @@ awesomebar:
...
+    description: |
+      The search term suggestion in awesomebar was clicked.
+    bugs:
+      - https://github.com/mozilla-mobile/fenix/issues/25816
+    data_reviews:
+      - ???????
+    data_sensitivity:
+      - interaction
+    notification_emails:
+      - android-probes@mozilla.com
```
Comment (rocketsroger) [inline app/metrics.yaml:7668]: Is we're collecting this permanently, please add kbrosnan@mozilla.com as well as a contact.  Thanks

**Comment by rocketsroger** [comment]
> ### **Request for data collection review**
> 
>     1. What questions will you answer with this data?
>        To better understand the usefulness of awesomebar suggestions .
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?    
>        This will allow to refine Awesomebar suggestions.
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?    
>        No alternative is available to answer this question.
> 
>     4. Can current instrumentation answer these questions?    
>        No current instrumentation can answer this question.
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.
> 
> 
> Measurement Description 	Data Collection Category 	Tracking Bug #
> Awesomebar search term suggestion was clicked 	category 2 - interaction data 	#25816
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
>        https://github.com/mozilla-mobile/fenix/blob/master/docs/metrics.md
> 
>     7. How long will this data be collected? Choose one of the following:    
>        This will be set to never expire, similar to the other telemetry for similar search suggestion interactions.
> 
>     8. What populations will you measure?    
>        No filters - all locales, channels, etc.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?    
>        Default Glean SDK opt-out mechanism.
> 
>     10. Please provide a general description of how you will analyze this data.     
>         Glean / Amplitude
> 
>     11. Where do you intend to share the results of your analysis?     
>         Only on Glean, Amplitude and with mobile teams.
> 
>     12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
>              No third-party tool will use this data.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/apps/fenix )

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, through the "Send Usage Data" preference in the application settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

Permanent with kbrosnan@mozilla.com monitoring the data over time.

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction data

5) Is the data collection request for default-on or default-off?

default-on

6) Does the instrumentation include the addition of **any *new* identifiers**?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

## Result

data-review+

**Review by rocketsroger — APPROVED** [review]
Looks good.  Only question left is if we should just use `lateinit` for 'engineForSpeculativeConnects`.  Thanks

File: app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt [hunk 2]
```diff
@@ -44,14 +52,13 @@ class AwesomeBarView(
...
     val interactor: AwesomeBarInteractor,
     val view: AwesomeBarWrapper,
-    fromHomeFragment: Boolean,
+    private val fromHomeFragment: Boolean,
 ) {
-    private val sessionProvider: SessionSuggestionProvider
-    private val historyStorageProvider: HistoryStorageSuggestionProvider
-    private val combinedHistoryProvider: CombinedHistorySuggestionProvider
+    private var components: Components = activity.components
+    private var engineForSpeculativeConnects: Engine? = null
```
[the remark this answers is not in the document]
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt:56]: With this being initialized in the `init` block I can postpone the initialization without even needing the `lateinit` keyword. Will try that. Thanks!

image

**Comment by Mugurell** [comment]
`addPredefinedSearchEngineTest` failed because of `ava.lang.IllegalArgumentException: Can't scroll to index 5, it is out of bounds [0, 3)` .
May be an old intermittently failing test - https://github.com/mozilla-mobile/fenix/issues/24024.
Locally this test passes.
~Restarting the task.~ Thanks to the kind soul who did this already!

**Pull request merged**